### PR TITLE
Allow parent_id pattern matching in Storage.get_all.

### PR DESCRIPTION
**Pull request by Natim**
r? @leplatrem 

- [x] Add tests.
- [x] Add a changelog entry.

### CODE REVIEW

**Comment by Natim** [comment]
Apparently there are a lot of edge case to support for that.

File: CHANGELOG.rst [hunk 1]
```diff
@@ -6,7 +6,9 @@ This document describes changes between each past release.
 4.3.0 (unreleased)
 ------------------
 
-- Nothing changed yet.
+**New features**
+
+- Allow ``parent_id`` pattern matching in ``Storage.get_all``. (#821)
```
Comment (leplatrem) [inline CHANGELOG.rst]: nit: The storage backend now allows [...] in ``kinto.core.storage.get_all()``.

File: tests/core/test_storage.py [hunk 2]
```diff
@@ -342,6 +342,17 @@ def test_delete_raise_when_unknown(self):
...
+        self.create_record(parent_id='abc', collection_id='c')
+        record = self.create_record(parent_id='abc', collection_id='c')
+        self.storage.delete(object_id=record['id'], parent_id='abc', collection_id='c')
+        self.create_record(parent_id='efg', collection_id='c')
+
+        records, total_records = self.storage.get_all(parent_id='ab*', collection_id='c',
+                                                      include_deleted=True)
+        self.assertEquals(len(records), 2)
+        self.assertEquals(total_records, 1)
+
```
Comment (leplatrem) [inline tests/core/test_storage.py]: I would suggest to add a test without parent matching to make sure that we kept the strict comparison behaviour when two parent_ids are similar (eg. `parent_id: xabcx` should not be returned with `parent_id=abc`)

By the way, maybe we should have it for `delete_all()` also...

File: kinto/core/storage/memory.py [hunk 3]
```diff
@@ -230,11 +230,36 @@ def get_all(self, collection_id, parent_id, filters=None, sorting=None,
...
+        for pid, collections in by_parent_id.items():
+            if collection_id is not None:
+                collections = {collection_id: collections[collection_id]}
+            for collection, colrecords in collections.items():
+                for r in colrecords.values():
+                    records.append(r)
+
+        records, count = self.extract_record_set(records,
+                                                 filters, None,
+                                                 id_field, deleted_field)
```
Comment (leplatrem) [inline kinto/core/storage/memory.py]: nit: since this method has many parameters, I suggest that we provide the kwargs names

File: kinto/core/storage/memory.py [hunk 4]
```diff
@@ -230,11 +230,36 @@ def get_all(self, collection_id, parent_id, filters=None, sorting=None,
...
+            by_parent_id = {pid: collections
+                            for pid, collections in self._cemetery.items()
+                            if parent_id_match.match(pid)}
+
+            for pid, collections in by_parent_id.items():
+                if collection_id is not None:
+                    collections = {collection_id: collections[collection_id]}
+                for collection, colrecords in collections.items():
+                    for r in colrecords.values():
+                        deleted.append(r)
```
Comment (leplatrem) [inline kinto/core/storage/memory.py]: nit: those two blocks of code are very similar. Maybe we could reuse a small helper function shared with `delete_all()` ?

File: tests/core/test_storage.py (hunk 2 above)
Comment (Natim) [inline tests/core/test_storage.py]: You are right, this tests fails with Postgresql :(

File: kinto/core/storage/memory.py (hunk 3 above)
Comment (Natim) [inline kinto/core/storage/memory.py]: It doesn't really improve readability because parameters have got the same name as the variable that contains the value.

`(records=records, filters=filters, sorting=None, id_field=id_field, deleted_field=deleted_field)`
